Fault Tolerance and tests for MPI

dmakoviichuk-tt opened the pull request:
### Problem description
* C MPI api doesn't throw exceptions. 
* Current infra doesn't support multihosts tests.
* Default openmpi doesn't support User Level Failure Mitigation (ULFM)

### What's changed
* Added compilation of the latest openmpi with the latest feature
* Add wrapper for all mpi calls to throw exceptions
* Add new multihost folder to tests
* Add baseline ULFM test

Please look at readme file: https://github.com/tenstorrent/tt-metal/blob/1774dd6e5a8167b6d03c5f9ab12da8067ab4c022/tests/multihost/README.md

https://github.com/tenstorrent/tt-metal/actions/runs/15145243109
### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [x] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [x] New/Existing tests provide coverage for changes

dmakoviichuk-tt committed:
Initial commit

dmakoviichuk-tt committed:
updated cmake files

dmakoviichuk-tt committed:
a

dmakoviichuk-tt committed:
now compiles

dmakoviichuk-tt committed:
updated sh

dmakoviichuk-tt committed:
removed comments

dmakoviichuk-tt committed:
fixed destructor

dmakoviichuk-tt committed:
fixed boost mpi

dmakoviichuk-tt committed:
updated cmake

dmakoviichuk-tt committed:
Merge branch 'main' into DM/mpi_support_metal

dmakoviichuk-tt committed:
temporary removed ensureboosttarget

dmakoviichuk-tt committed:
Merge branch 'DM/mpi_support_metal' of https://github.com/tenstorrent/tt-metal into DM/mpi_support_metal

afuller-TT committed:
load non-header components from system Boost, too

dmakoviichuk-tt committed:
distributed context now uses c-mpi back

dmakoviichuk-tt committed:
merged cmake

dmakoviichuk-tt committed:
updated third_party

dmakoviichuk-tt committed:
space

dmakoviichuk-tt committed:
updated license

dmakoviichuk-tt committed:
fixed codestyle

dmakoviichuk-tt committed:
fixed space

dmakoviichuk-tt committed:
Update tt_metal/distributed/CMakeLists.txt

Co-authored-by: Andrew Fuller <afuller@tenstorrent.com>

dmakoviichuk-tt committed:
added exception

dmakoviichuk-tt committed:
added exceptions

dmakoviichuk-tt committed:
added exceptions

dmakoviichuk-tt committed:
initial commit

dmakoviichuk-tt committed:
Added readme

omilyutin-tt commented on tests/multihost/single_host_mp_tests/main.cpp:
How does this get used? `main` is used along side what gtest defines for us?

dmakoviichuk-tt replied:
GTest::gmock_main if it is linked you have google main.
I need to make sure I don't link this.

omilyutin-tt commented on tt_metal/distributed/mpi_distributed_context.cpp:
```diff
@@ -32,36 +33,63 @@ inline void check_size_fits_int(std::size_t n) {
+    }
+}
+
+#define MPI_CHECK(call) mpi_check((call), #call)
```
Wondering why macro is neede here

dmakoviichuk-tt replied:
easiest way to wrap all call in one line ang get a function name. Also I can add file and line if needed.
If I go templates way (which I tried):
ill need to find a way to get function name provided to the template (will require demangle and typeid on template paramter).
Impossible to get __LINE__ and __FILE__ in easy way (std::source_location not supported in c++20).
As result my true c++ approach took 100+ lines vs 1 :)

cfjchu commented on tt_metal/distributed/mpi_distributed_context.cpp:
```diff
@@ -241,52 +278,58 @@ void MPIContext::scan(tt::stl::Span<std::byte> send_buf,
 
     std::vector<int> int_ranks(ranks.size());
     std::transform(ranks.begin(), ranks.end(), int_ranks.begin(),
                    [](Rank r) { return *r; });
 
-    MPI_Group_incl(world_grp, static_cast<int>(int_ranks.size()),
-                   int_ranks.data(), &sub_grp);
+    MPI_CHECK(MPI_Group_incl(world_grp, static_cast<int>(int_ranks.size()), int_ranks.data(), &sub_grp));
 
     MPI_Comm sub_comm;
-    MPI_Comm_create_group(comm_, sub_grp, 0 /*tag*/, &sub_comm);
-    MPI_Group_free(&sub_grp);
-    MPI_Group_free(&world_grp);
+    MPI_CHECK(MPI_Comm_create_group(comm_, sub_grp, 0 /*tag*/, &sub_comm));
+
+    // TODO: unsafe free calls. now mpi can throw
+    MPI_CHECK(MPI_Group_free(&sub_grp));
+    MPI_CHECK(MPI_Group_free(&world_grp));
 
     return std::make_shared<MPIContext>(sub_comm);
```
since `MPI_CHECK` returns exceptions now, why not wrap this code block directly w/ try-catch and free only when it's safe

dmakoviichuk-tt replied:
I have 2 free functions. So it will be two blocks or soemthing like this :( I was going to add a simple scope exit each time after create call.

cfjchu commented on tt_metal/distributed/mpi_distributed_context.cpp:
```diff
@@ -74,7 +102,10 @@ static void init_env(int& argc, char**& argv)
     if (!initialized) {
         int provided = 0;
-        MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided);
+        int rc = MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided);
+        if (rc != MPI_SUCCESS) {
+            TT_THROW("MPI_Init_thread failed");
+        }
```
why not use `MPI_CHECK` here too

dmakoviichuk-tt commented on tt_metal/distributed/mpi_distributed_context.cpp:
```diff
@@ -22,6 +19,10 @@ inline MPI_Op reduce_to_mpi(ReduceOp op) {
+        case ReduceOp::BAND: return MPI_BAND;
+        case ReduceOp::BOR: return MPI_BOR;
     }
     return MPI_SUM; // default
```
we should never get here. but throw is ok too.

dmakoviichuk-tt replied:
ah I think I don't want to throw in constexpr.

dmakoviichuk-tt commented on tests/multihost/single_host_mp_tests/main.cpp:
```diff
@@ -0,0 +1,22 @@
+    int local_rc = RUN_ALL_TESTS();
+
+    // Propagate the worst return code to all ranks
+    int global_rc = local_rc;
```
yes :) I decided to move all this sthings to the common library but didn't implement it yet.

cfjchu commented on tt_metal/distributed/mpi_distributed_context.cpp:
```diff
@@ -74,7 +102,10 @@ static void init_env(int& argc, char**& argv)
     if (!initialized) {
         int provided = 0;
-        MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided);
+        int rc = MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided);
```
I just noticed `MPI_THREAD_MULTIPLE`... are the DistributedContext methods thread-safe?

dmakoviichuk-tt replied:
Yes, you can call it from different threads and it should work. But initialization is not (AFAIK).

omilyutin-tt replied:
`std::call_once` might be relevant here

dmakoviichuk-tt committed:
moved multihost tests to the tt_metal

omilyutin-tt commented on tests/tt_metal/multihost/single_host_mp_tests/test_send_recv.cpp:
```diff
@@ -0,0 +1,34 @@
+
+TEST(DistributedContextTest, TestSendRecv) {
+    // assuming context is already initialized in main with argc, argv
+    // in this case we will jsut get a world context
```
```suggestion
    // in this case we will just get a world context
```

dmakoviichuk-tt committed:
Update tests/tt_metal/multihost/single_host_mp_tests/test_send_recv.cpp

Co-authored-by: Oleg Milyutin <omilyutin@tenstorrent.com>

dmakoviichuk-tt committed:
added reduce fucntions with type deductions

dmakoviichuk-tt committed:
updated tests

dmakoviichuk-tt committed:
tests updated

dmakoviichuk-tt committed:
updated with latest

dmakoviichuk-tt committed:
updated tests

ayerofieiev-tt commented on tests/tt_metal/multihost/common/multihost_asserts.hpp:
```diff
@@ -0,0 +1,112 @@
+// ----------------------------------------------------------------------
+//  multihost test main function
+// ----------------------------------------------------------------------
+inline int multihost_main(int argc, char** argv) {
```
please not in multihost_asserts.hpp 🥹

ayerofieiev-tt commented on tests/tt_metal/multihost/fault_tolerance_tests/ulfm_tests.cpp:
```diff
@@ -0,0 +1,56 @@
+    //----------------------------------------------------------------------
+    // 0 · Create world communicator and install MPI_ERRORS_RETURN
+    //----------------------------------------------------------------------
+    auto ctx = DistributedContext::create(0, nullptr);
```
(nit) can you give 0 and nullptr a comment or put into a variable with a clear name

ayerofieiev-tt commented on tests/tt_metal/multihost/fault_tolerance_tests/ulfm_tests.cpp:
```diff
@@ -0,0 +1,56 @@
+    //----------------------------------------------------------------------
+    auto ctx = DistributedContext::create(0, nullptr);
+
+    const int world = *ctx->size();
```
can't be?
```suggestion
    const int world = ctx->size();
```

tt-rkim commented on tests/tt_metal/multihost/multi_host_tests/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+# Smoke tests (fast, not necessarily thorough)
```
Can we add as part of the debian to the merge queue suite then? That would be great!

dmakoviichuk-tt replied:
I didn't plan to turn on these tests in the pr.
For the multi_host we don't have required hardware.
I was planning to turn them on in the next pr.

tt-rkim replied:
Can we skip if multi host isn't detected?

Are these basic smoke tests?

Denys88 replied:
Yeah we can skip. But still I am planning to work on it next pr.
Current goal is to unlock multihost work.

tt-rkim commented on tests/tt_metal/multihost/multi_host_tests/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+# Smoke tests (fast, not necessarily thorough)
+add_library(mullti_host OBJECT)
```
Seems like you have some spelling errors, here, boss
```suggestion
add_library(multi_host OBJECT)
```

dmakoviichuk-tt replied:
ouch! thanks!

tt-rkim commented on third_party/CMakeLists.txt:
```diff
@@ -265,3 +265,47 @@ if(simd-everywhere_ADDED)
+############################################################################################################################
+# ULFM is a fork of OpenMPI with support for fault tolerance
+
+CPMAddPackage(
```
@afuller-TT @blozano-tt Anything special we'd need to here for packaging?

tt-rkim replied:
Anything special we'd have to do here for packaging?

Do we need to expose any filesets to CMake to ensure things here are included? @afuller-TT

afuller-TT commented on tests/tt_metal/multihost/fault_tolerance_tests/main.cpp:
```diff
@@ -0,0 +1,9 @@
+#include <gtest/gtest.h>
+#include "common/multihost_asserts.hpp"
+#include <fmt/format.h>
+int main(int argc, char** argv) { return multihost::common::multihost_main(argc, argv); }
```
Why do we need our own main?  Is GMock's main not good enough for some reason?

dmakoviichuk-tt replied:
Yes, not enough. For the multihost usecase we need to add 2 things:
1) initialize mpi from argc, argv.
2) all_reduce results

afuller-TT commented on tt_metal/distributed/CMakeLists.txt:
```diff
@@ -15,8 +15,33 @@ set(DISTRIBUTED_SRC
 add_library(distributed OBJECT ${DISTRIBUTED_SRC})
+add_dependencies(distributed openmpi_ulfm_build)
+
+set(ULFM_PREFIX ${CMAKE_BINARY_DIR}/openmpi-ulfm-install)
```
This target seems tied at the hip to ULFM_MPI.  Better to move it to where we're handling ULFM_MPI so we can reason through them both together.

dmakoviichuk-tt replied:
I am not sure It might didn't work but Ill try again.
I must wait for external job to finish before using it: 
thats why we have this dependency.
add_dependencies(distributed openmpi_ulfm_build)

I'd like to move all this things:
```
set(ULFM_PREFIX ${CMAKE_BINARY_DIR}/openmpi-ulfm-install)

add_library(OpenMPI::MPI SHARED IMPORTED)

# Path to the actual library file, not the directory
find_library(ULFM_MPI_LIB NAMES mpi PATHS ${ULFM_PREFIX}/lib NO_DEFAULT_PATH)

set_target_properties(
    OpenMPI::MPI
    PROPERTIES
        IMPORTED_LOCATION
            ${ULFM_MPI_LIB}
        INTERFACE_INCLUDE_DIRECTORIES
            ${ULFM_PREFIX}/include
        INTERFACE_LINK_LIBRARIES
            ${ULFM_PREFIX}/lib/libmpi.so
        INTERFACE_C_INCLUDE_DIRECTORIES
            ${ULFM_PREFIX}/include
        INTERFACE_CXX_INCLUDE_DIRECTORIES
            ${ULFM_PREFIX}/include
        INTERFACE_SYSTEM_INCLUDE_DIRECTORIES
            ${ULFM_PREFIX}/include
)
```
back but for some reason dependecies become broken because I need to wait openmpi_ulfm_build before I start using OpenMPI::MPI

dmakoviichuk-tt replied:
I think I fixed it. Now looks better.

afuller-TT commented on tt_metal/distributed/CMakeLists.txt:
```diff
@@ -15,8 +15,33 @@ set(DISTRIBUTED_SRC
+        IMPORTED_LOCATION
+            ${ULFM_MPI_LIB}
+        INTERFACE_INCLUDE_DIRECTORIES
+            ${ULFM_PREFIX}/include
```
The build is failing because it claims this dir does not exist.  It works for you locally?  Is that leftover from a previous run, perhaps?

dmakoviichuk-tt replied:
maybe, I'll make a clean run locally once again.

dmakoviichuk-tt committed:
finally it is stable

dmakoviichuk-tt committed:
fixed cmake

dmakoviichuk-tt committed:
removed file

dmakoviichuk-tt committed:
ompi build requires it

dmakoviichuk-tt committed:
working solution

dmakoviichuk-tt committed:
added libs to build mpi

dmakoviichuk-tt committed:
remvoed as_bytes

dmakoviichuk-tt committed:
Merge branch 'main' into DM/mpi_safety

dmakoviichuk-tt committed:
fixed Oleg

dmakoviichuk-tt committed:
trying to fix it

dmakoviichuk-tt committed:
Another option how to add mpi with ULFM (#22129)

Please look at the main pr for details:
https://github.com/tenstorrent/tt-metal/pull/21870
### Problem description
Add_ExternalProject is too complex

### What's changed
Implemented MPI 5.0.7 installation as a part of install_dependencies.sh
script

dmakoviichuk-tt committed:
Merge branch 'main' into DM/mpi_safety

ttmchiou commented on dockerfile/Dockerfile.basic-dev:
```diff
@@ -17,3 +17,15 @@ RUN apt-get update && apt-get install -y \
+    apt-get install -y -f \
+        wget ca-certificates && \
+    TMP_DIR="$(mktemp -d)" && \
+    DEB_URL="https://github.com/dmakoviichuk-tt/mpi-ulfm/releases/download/v5.0.7-ulfm/openmpi-ulfm_5.0.7-1_amd64.deb" && \
```
can we put this mpi-ulfm into the TT org instead of your own personal repo?

dmakoviichuk-tt replied:
I will do it in a future.
Last time I tried to create a repo it took 2 months of waiting for approval.

dmakoviichuk-tt committed:
passes all the tests

blozano-tt commented on tests/tt_metal/multihost/README.md:
````diff
@@ -0,0 +1,81 @@
+## Building
+
+```bash
+CMake uses external project to build mpi with ULFM. Make sure you don't use a system mpirun
````
Is this still true?

blozano-tt commented:
https://github.com/tenstorrent/tt-metal/actions/runs/15129013327

Running APC to sanity check ^

dmakoviichuk-tt committed:
fixed tttron

dmakoviichuk-tt committed:
added include

dmakoviichuk-tt committed:
fixed thread in another test

dmakoviichuk-tt committed:
added multiple block test

blozano-tt reviewed (approved):
Please resolve any open conversations before merge.

dmakoviichuk-tt committed:
updated readme

dmakoviichuk-tt committed:
Merge branch 'main' into DM/mpi_safety

dmakoviichuk-tt committed:
updated install dependencies

dmakoviichuk-tt committed:
Merge branch 'DM/mpi_safety' of https://github.com/tenstorrent/tt-metal into DM/mpi_safety
